docs/source/compilation-databases.rst

```diff
@@ -0,0 +1,126 @@
+Compilation Databases
+=====================
+
+Before it can analyze a code base, CBI needs to know how each source file is
+compiled. Just like a compiler, CBI requires a full list of include paths,
+macro definitions and other options in order to identify which code is used
+by each platform. Rather than require all of this information to be specified
+manually, CBI reads it from a `compilation database`_.
+
+
+Generating a Compilation Database
+#################################
+
+Since our sample code base is already set up with a ``CMakeLists.txt`` file, we
+can ask CMake to generate the compilation database for us with the
+:code:`CMAKE_EXPORT_COMPILE_COMMANDS` option:
+
+.. code-block:: cmake
+    :emphasize-lines: 4
+
+    cmake_minimum_required(VERSION 3.5)
+    project(tutorial)
+
+    set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+
+    set(SOURCES main.cpp third-party/library.cpp)
+
+    option(GPU_OFFLOAD "Enable GPU offload." OFF)
+    if (GPU_OFFLOAD)
+        add_definitions("-D GPU_OFFLOAD=1")
+        list(APPEND SOURCES gpu/foo.cpp)
+    else()
+        list(APPEND SOURCES cpu/foo.cpp)
+    endif()
+
+    add_executable(tutorial ${SOURCES})
+
+.. important::
+    For projects that don't use CMake, we can use `Bear`_ to intercept the
+    commands generated by other build systems (such as GNU makefiles). Other
+    build systems and tools that produce compilation databases should also be
+    compatible.
+
+.. _`compilation database`: https://clang.llvm.org/docs/JSONCompilationDatabase.html
+.. _`Bear`: https://github.com/rizsotto/Bear
+
+
+CPU Compilation Commands
+------------------------
+
+Let's start by running CMake without the :code:`GPU_OFFLOAD` option enabled, to
+obtain a compilation database for the CPU:
+
+.. code :: sh
+
+    $ mkdir build-cpu
+    $ cmake ../
+    $ ls
+
+    CMakeCache.txt  CMakeFiles  Makefile  cmake_install.cmake  compile_commands.json
+
+This :code:`compile_commands.json` file includes all the commands required to
+build the code, corresponding to the commands that would be executed if we were
+to actually run :code:`make`.
+
+.. attention::
+    CMake generates compilation databases when the ``cmake`` command is
+    executed, allowing us to generate compilation databases without also
+    building the application. Other tools (like Bear) may require a build.
+
+In this case, it contains::
+
+    [
```

Comment:
Is this meant to be in a code block?

Reply:
This is RST magic.  If you have two colons `::` then it gives the block below special handling.  I'll check whether explicitly calling this out as JSON renders things differently.

Reply:
OK, explicitly specifying JSON is better.  Good catch!  Fixed in 123d24efd632d2f5155281c2b8c6a4715e3c12ad.